### PR TITLE
hash:feature - new hash format and handle the future depreciations

### PR DESCRIPTION
**Pull request by nathanmartinszup**
Currently cli generates very volatile hashes, any commit, description
changes will change the hash, wich is a problem to the vulnerabilities
management. This pull request changes the way that we generate the hash
to use only the code, line and file, this way we expect that the hash
will be more estable. In this pull request we also added a tratative to
handle with the future depreciation of the old hashes that we still
consider valid. So all users should update the oldated hashes, the cli
is going to show a warning with the necessary hashes to update and the
new ones.

Signed-off-by: Nathan Martins <nathan.martins@zup.com.br>

**- What I did**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: internal/controllers/analyzer/analyzer.go [hunk 1]
```diff
@@ -370,3 +392,37 @@ func (a *Analyzer) removeVulnerabilitiesByTypes() *analysis.Analysis {
 
 	return a.analysis
 }
+
+// setUpdateHashWarnings this func checks for hashes generated in older formats but that are still valid. If one of
```
Comment (matheusalcantarazup) [inline internal/controllers/analyzer/analyzer.go:396]: I think that "this func" is not needed.

File: internal/controllers/analyzer/analyzer.go [hunk 2]
```diff
@@ -370,3 +392,37 @@ func (a *Analyzer) removeVulnerabilitiesByTypes() *analysis.Analysis {
...
+// these hashes are found, a warning will be showed informing the user to update the outdated hash.
+// TODO: Remove setUpdateHashWarnings before release v2.10.0
+// nolint
+func (a *Analyzer) setUpdateHashWarnings() {
+	isPrintDepreciationMsg := true
+
+	for vulnIndex := range a.analysis.AnalysisVulnerabilities {
+		for _, deprecatedHash := range a.analysis.AnalysisVulnerabilities[vulnIndex].Vulnerability.DeprecatedHashes {
```
Comment (matheusalcantarazup) [inline internal/controllers/analyzer/analyzer.go:404]: I think that would be good create a temp variable that holds `a.analysis.AnalysisVulnerabilities[vulnIndex]` value

File: internal/controllers/printresults/print_results.go [hunk 3]
```diff
@@ -433,3 +434,10 @@ func (pr *PrintResults) createTxtOutputFile() error {
 
 	return file.CreateAndWriteFile(pr.textOutput, pr.config.JSONOutputFilePath)
 }
+
+// printWarnings function that will print all necessary warnings in the end of the analysis
```
Comment (matheusalcantarazup) [inline internal/controllers/printresults/print_results.go:438]: "function that will" is not necessary.

File: internal/utils/vuln_hash/vuln_hash.go [hunk 4]
```diff
@@ -28,21 +28,37 @@ import (
...
+	// It generates the hash without the commitEmail and description information, avoiding frequent hash changes
+	// every time that this data changes. Below two more hashes are generated, these contain two old versions of
+	// hashes that are still valid. After the depreciation, only this hash generated with row, column and code
+	// will be considered valid.
 	vuln.VulnHash = crypto.GenerateSHA256(
 		toOneLine(vuln.Code),
 		vuln.Line,
-		vuln.Details,
+
+		// The filepath should be relative to avoid generating different hashes between environments
```
Comment (matheusalcantarazup) [inline internal/utils/vuln_hash/vuln_hash.go:39]: I think that this comment should be added on Bind function docs?

```go
// Bind create a sha256 hash of the vulnerability using the vulnerability
// code, line and file. The file path should be relative to avoid generating 
// different hashes between environments.
//
// Bind automatically set the Vulnerability.VulnHash field with the generated
// vulnerability hash.
```

What do you think?

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]

File: internal/utils/vuln_hash/vuln_hash.go (hunk 4 above)
Comment (matheusalcantarazup) [reply internal/utils/vuln_hash/vuln_hash.go:39]: I noticed that Bind already have a godoc, but I think that don't have such information like the example above.

**Review by wiliansilvazup — APPROVED** [review]

File: internal/controllers/analyzer/analyzer.go [hunk 5]
```diff
@@ -159,30 +161,45 @@ func (a *Analyzer) formatAnalysisToSendToAPI() {
...
-		for idx := range a.analysis.AnalysisVulnerabilities {
-			vulnHash := a.analysis.AnalysisVulnerabilities[idx].Vulnerability.VulnHash
-			vulnHashInvalid := a.analysis.AnalysisVulnerabilities[idx].Vulnerability.VulnHashInvalid
-			if hash == vulnHash || hash == vulnHashInvalid {
-				existing = true
+// logWarnIfHashDontExistsInConfig logs a warning if the one of the config hashes don't exist in the analysis.
+// nolint
+func (a *Analyzer) logWarnIfHashDontExistsInConfig(configHashes []string) {
+	for _, configHash := range configHashes {
+		isDontExists := true
```
[the remark this answers is not in the document]
Comment (nathanmartinszup) [reply internal/controllers/analyzer/analyzer.go:168]: I don't think I understand the suggestion.

[the remark this answers is not in the document]
Comment (iancardosozup) [reply internal/controllers/analyzer/analyzer.go:168]: change isDontExists to exists,

[the remark this answers is not in the document]
Comment (nathanmartinszup) [reply internal/controllers/analyzer/analyzer.go:168]: But the function only logs if it does not exist, it will stop making sense.

File: internal/controllers/analyzer/analyzer.go [hunk 6]
```diff
@@ -372,3 +394,38 @@ func (a *Analyzer) removeVulnerabilitiesByTypes() *analysis.Analysis {
...
+// these hashes are found, a warning will be showed informing the user to update the outdated hash.
+// TODO: Remove setUpdateHashWarnings before release v2.10.0
+// nolint
+func (a *Analyzer) setUpdateHashWarnings() {
+	isPrintDepreciationMsg := true
+
+	vulnerabilities := a.analysis.AnalysisVulnerabilities
+	for vulnIndex := range vulnerabilities {
+		for _, deprecatedHash := range vulnerabilities[vulnIndex].Vulnerability.DeprecatedHashes {
+			for _, configHash := range a.getAllConfigHashes() {
```
Comment (matheusalcantarazup) [inline internal/controllers/analyzer/analyzer.go:408]: I think that would be good to create a variable with the results of `a.getAllConfigHashes()` to avoid recompute this value very time.

File: internal/controllers/analyzer/analyzer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (iancardosozup) [reply internal/controllers/analyzer/analyzer.go:168]: for me a variable named `isDontExists` is a duplicate negative because is just a negation of an `exists` variable, we'll have to invert the if logics to make the rest of the code right but it will be more readable
```go
exists := false
isDontExists := true
fmt.Println(!exists == isDontExists) // will print true
```
do you agree `isDontExists` make the code more confusing?

File: internal/services/formatters/csharp/dotnet_cli/formatter.go [hunk 7]
```diff
@@ -154,22 +154,31 @@ func (f *Formatter) parseFieldByIndex(index int, fieldValue string, dependency *
...
 		Details:      dependency.getDescription(),
 		Code:         code,
-		File:         f.removeHorusecFolder(filePath),
+		File:         relFilePath,
 		Line:         line,
 		Severity:     dependency.getSeverity(),
 	}
-	return f.SetCommitAuthor(vulnHash.Bind(vuln)), nil
+
+	vuln.DeprecatedHashes = f.getDeprecatedHashes(absFilePath, *vuln)
```
Comment (matheusalcantarazup) [inline internal/services/formatters/csharp/dotnet_cli/formatter.go:179]: Why do we need this if vulnhash.Bind already set the deprecated hashes?

File: internal/services/formatters/csharp/dotnet_cli/formatter.go [hunk 8]
```diff
@@ -154,22 +154,31 @@ func (f *Formatter) parseFieldByIndex(index int, fieldValue string, dependency *
...
+	code, absFilePath, line, err := f.getCodeFilePathAndLine(dependency, projectSubPath)
 	if err != nil {
 		return nil, err
 	}
+
+	relFilePath, err := filepath.Rel(f.GetConfigProjectPath(), absFilePath)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (matheusalcantarazup) [inline internal/services/formatters/csharp/dotnet_cli/formatter.go:166]: Why this change is needed?

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]
Its not clear to me why these changes in some formatters are necessary.

File: internal/controllers/analyzer/analyzer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathanmartinszup) [reply internal/controllers/analyzer/analyzer.go:168]: I was thinking it was related to the function name, with the example it became more clear what you meant.

File: internal/services/formatters/csharp/dotnet_cli/formatter.go (hunk 7 above)
Comment (nathanmartinszup) [reply internal/services/formatters/csharp/dotnet_cli/formatter.go:179]: The hash was being generated with an absolute path, with the change to the relative one, it was necessary to add the hashes generated with the absolute path as well.


The getDeprecatedHashes function has a comment that explains this.

File: internal/services/formatters/csharp/dotnet_cli/formatter.go (hunk 8 above)
Comment (nathanmartinszup) [reply internal/services/formatters/csharp/dotnet_cli/formatter.go:166]: Same as https://github.com/ZupIT/horusec/pull/996#discussion_r814818220

File: internal/services/formatters/csharp/dotnet_cli/formatter.go (hunk 7 above)
Comment (matheusalcantarazup) [reply internal/services/formatters/csharp/dotnet_cli/formatter.go:179]: Make totally sense, I was not understanding the needed to generate the hash twice.

File: internal/controllers/analyzer/analyzer.go (hunk 6 above)
Comment (nathanmartinszup) [reply internal/controllers/analyzer/analyzer.go]: Well observed, done!

File: internal/controllers/analyzer/analyzer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathanmartinszup) [reply internal/controllers/analyzer/analyzer.go]: Done!
